[nativert] Move Placement to pytorch core

yushangdi opened the pull request:
Summary:
Move Placement to pytorch core.

Using `torch::nativert::isSameDevice` explicitly in code to avoid confusion with the `isSameDevice` in torch namespace.

Test Plan:
```
buck run fbcode//mode/dev-nosan  //caffe2/test/cpp/nativert:placement_test

./bin/test_nativert 
```

OSS and internal CI

Differential Revision: D74190745

Skylion007 commented on torch/nativert/executor/Placement.cpp:
```diff
@@ -0,0 +1,91 @@
+std::ostream& operator<<(std::ostream& os, const Placement& placement) {
+  std::map<std::string, c10::Device> keys;
+  for (const auto& pair : placement.deviceMap_) {
+    keys.insert({pair.first.str(), pair.first});
```
```suggestion
    keys.try_emplace(pair.first.str(), pair.first);

Skylion007 commented on torch/nativert/executor/Placement.cpp:
```diff
@@ -0,0 +1,91 @@
+    keys.insert({pair.first.str(), pair.first});
+  }
+
+  bool first = true;
+  auto checkComma = [&]() {
+    if (!first) {
+      os << ",";
+    }
+    first = false;
+  };
+
+  os << "";
+  for (const auto& pair : keys) {
+    checkComma();
+    const auto& key = pair.second;
+    const auto& value = placement.deviceMap_.at(key);
+    os << pair.first << "|" << value.str();
+  }
+
+  if (placement.defaultDevice_.has_value()) {
+    checkComma();
+    os << "|" << placement.defaultDevice_.value().str();
+  }
+  return os;
```
```suggestion
  std::vector<std::string> entries;
  entries.reserve(keys.size() + 1);
  for (const auto& pair : keys) {
    const auto& key = pair.second;
    const auto& value = placement.deviceMap_.at(key);
    entries.emplace_back(fmt::format("{}|{}", pair.first, value.str()));
  }

  if (placement.defaultDevice_.has_value()) {
    entries.emplace_back(fmt::format("|{}", placement.defaultDevice_->str()));
  }

  fmt::print(os, "{}", fmt::join(entries, ","));
  return os;

Skylion007 commented on torch/nativert/executor/Placement.cpp:
```diff
@@ -0,0 +1,91 @@
+#include <torch/nativert/executor/Placement.h>
+
+#include <ostream>
```
```suggestion
#include <ostream>
#include <fmt/ostream.h>
#include <fmt/ranges.h>
```

Skylion007 commented on test/cpp/nativert/CMakeLists.txt:
```diff
@@ -3,7 +3,9 @@ set(NATIVERT_TEST_ROOT ${TORCH_ROOT}/test/cpp/nativert)
   ${NATIVERT_TEST_ROOT}/test_tensor_meta.cpp
+  ${NATIVERT_TEST_ROOT}/test_placement.cpp
   ${TORCH_ROOT}/torch/nativert/graph/TensorMeta.cpp
+  ${TORCH_ROOT}/torch/nativert/executor/Placement.cpp
```
Seems like you are missing includes here?

zhxchen17 replied:
@yushangdi I don't think you need to include this here because you already marked Placement as TORCH_API so the test binary can find the implementation from libtorch directly.

Alternatively you can remove TORCH_API marker and still keep this source file here. But it's strange to have both TORCH_API and this being builtin into the test binary directly (which may cause ODR violation)

yushangdi replied:
> @yushangdi I don't think you need to include this here because you already marked Placement as TORCH_API so the test binary can find the implementation from libtorch directly.
> 
> Alternatively you can remove TORCH_API marker and still keep this source file here. But it's strange to have both TORCH_API and this being builtin into the test binary directly (which may cause ODR violation)

@zhxchen17 Sure. I removed the dependency.

albanD commented on torch/nativert/executor/Placement.h:
```diff
@@ -0,0 +1,29 @@
+#include <optional>
+#include <unordered_map>
+
+namespace torch::nativert {
```
Can you please add documentation for all of these as it is not very clear what the intended use of these is.

yushangdi replied:
> Can you please add documentation for all of these as it is not very clear what the intended use of these is.

@albanD added now!

albanD commented on torch/nativert/executor/Placement.h:
```diff
@@ -0,0 +1,60 @@
+ * @throws c10::Error If the device type is not CPU or CUDA.
+ */
+
+TORCH_API c10::Device normalizeDevice(const c10::Device& device);
```
I don't think this should be a public API.

albanD commented on torch/nativert/executor/Placement.h:
```diff
@@ -0,0 +1,60 @@
+namespace torch::nativert {
+
+/**
+ * @brief Normalizes a c10::Device by ensuring a canonical form.
```
nit: a "canonical form" for Device is not a thing within PyTorch. So you most likely want to use another word here.

albanD commented on torch/nativert/executor/Placement.h:
```diff
@@ -0,0 +1,60 @@
+TORCH_API c10::Device normalizeDevice(const c10::Device& device);
+
+/**
+ * Returns true if the two devices are the same and has the same device index (if cuda).
```
This is only with the above definition of device index not being important on CPU?

yushangdi replied:
yeah. and the device index is not compared in the implementation if the device is cpu.

albanD replied:
I would be curious why you had to build so much scaffolding to special case cpu here and not just use whatever index is used there (which is going to be 0 in ~100% of the cases).

yushangdi replied:
> I would be curious why you had to build so much scaffolding to special case cpu here and not just use whatever index is used there (which is going to be 0 in ~100% of the cases).

good question, I'm not sure, as this is mostly exiting code moved from internal. @SherlockNoMad Do you know why we need to special case for CPU here?

albanD commented on torch/nativert/executor/Placement.h:
```diff
@@ -0,0 +1,60 @@
+/**
+ * Returns true if the two devices are the same and has the same device index (if cuda).
+ */
+TORCH_API bool isSameDevice(
```
Shouldn't be public either.

albanD replied:
The semantic feels so specific and different from what the name suggests that I don't think we want to expose this to all OSS users.

yushangdi replied:
i see. removed now

albanD commented on torch/nativert/executor/Placement.h:
```diff
@@ -0,0 +1,60 @@
+ *
+ * The Placement class provides a way to map source devices to target devices.
+ * It supports both explicit per-device mappings and a default device fallback.
+ * This is used for device parameterization in Torch Native Runtime, and to support
```
What does "device parametrization" means?

albanD replied:
Since this is only to be used with the nativert API, I would say it the way you said it above: this is the argument taken in nativert to map from model artifact device to the device it should run on.

albanD commented on torch/nativert/executor/Placement.h:
```diff
@@ -0,0 +1,60 @@
+ * This is used for device parameterization in Torch Native Runtime, and to support
+ * execution with various backends, or with various device placement.
+ */
+struct TORCH_API Placement {
```
Do you need this to be public?

zhxchen17 replied:
@albanD Yeah I think this is part of the nativert API. I agree all the other things might not need to be TORCH_API, but Placement is something user provide to us to define where to execute the model, according to the RFC https://github.com/pytorch/rfcs/blob/1175e55852c9a95b3131607a04d184055e37ebde/RFC-0043-torch-native-runtime.md#proposed-implementation

albanD replied:
Sounds good!

yushangdi replied:
Users will might need to construct their Placement and use it as an input to the runtime, so I think we might need to make it public?

yushangdi replied:
@albanD I think this Placement needs to be public because people might need to construct their own Placement object and pass it into the runtime.

zhxchen17 commented on torch/nativert/executor/Placement.cpp:
```diff
@@ -0,0 +1,85 @@
+#include <torch/nativert/executor/Placement.h>
```
@cyyever We've been in talk with @albanD  And @nikithamalgifb and our conclusion was that nativert should go into its own directory and not csrc/ .

zhxchen17 replied:
Updated RFC: https://github.com/pytorch/rfcs/blob/fc0c9c2fee82828e0bc8eb33a58c05acfabce3a9/RFC-0043-torch-native-runtime.md#proposed-implementation

yushangdi replied:
@Skylion007 We're putting the Native Runtime related code in nativert according to the  RFC: https://github.com/pytorch/rfcs/blob/fc0c9c2fee82828e0bc8eb33a58c05acfabce3a9/RFC-0043-torch-native-runtime.md#proposed-implementation

swolchok commented on torch/nativert/executor/Placement.cpp:
```diff
@@ -0,0 +1,55 @@
+namespace torch::nativert {
+
+std::ostream& operator<<(std::ostream& os, const Placement& placement) {
+  std::map<std::string, c10::Device> keys;
+  for (const auto& pair : placement.deviceMap_) {
+    keys.try_emplace(pair.first.str(), pair.first);
+  }
```
why the std::map? if you just want sorted iteration, make a vector and sort it.

yushangdi replied:
> why the std::map? if you just want sorted iteration, make a vector and sort it.

sure, changing to the following:

```
  std::vector<std::pair<std::string, c10::Device>> sorted_keys;
  for (const auto& pair : placement.deviceMap_) {
    sorted_keys.emplace_back(pair.first.str(), pair.first);
  }
  std::sort(sorted_keys.begin(), sorted_keys.end(),
            [](const auto& a, const auto& b) {
              return a.first < b.first;
            });

  bool first = true;
  for (const auto& pair : sorted_keys) {
...
```

swolchok reviewed (approved):
seems fine

cyyever commented on torch/nativert/executor/PlacementUtils.cpp:
```diff
@@ -0,0 +1,37 @@
+  }
+}
+
+bool isSameDevice(const c10::Device& a, const c10::Device& b) {
```
Here only `cpu` and `cuda` are mentioned, how about other devices? Should we take this opportunity to enhance it?

yushangdi replied:
> Here only `cpu` and `cuda` are mentioned, how about other devices? Should we take this opportunity to enhance it?

@cyyever  I see in the original code D56254093 that other devices are intentionally left behind until we introduce their backend.

cyyever commented:
Is it possible to not only move code to a new place but also change them to support other kinds of device?

yushangdi commented:
> Is it possible to not only move code to a new place but also change them to support other kinds of device?

@cyyever we're focusing on moving code so we might want to avoid behavior changes when possible. If we want to support more devices, I can make a follow up PR.

yushangdi commented:
@Skylion007 No merge conflict now! :)

Skylion007 commented:
Make sure this is rebased so you get the clang-tidy checks.

yushangdi commented:
> Make sure this is rebased so you get the clang-tidy checks.

@Skylion007 Do you mean the checks in https://github.com/pytorch/pytorch/pull/153164? This PR is already rebased on top of it, so it's getting those checks

yushangdi committed:
[nativert] Move Placement to pytorch core (#152953)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/152953

Move Placement to pytorch core.

Using `torch::nativert::isSameDevice` explicitly in code to avoid confusion with the `isSameDevice` in torch namespace.

Place non-public implementations in a separate cpp file to avoid duplicated symbol with libtorch.

Test Plan:
```
buck run fbcode//mode/dev-nosan  //caffe2/test/cpp/nativert:placement_test
```

OSS and internal CI

Reviewed By: zhxchen17

Differential Revision: D74190745
